### PR TITLE
Remove JSON.stringify/parse, addd onProcessStyle hook and more.

### PR DESCRIPTION
**Pull request by kof**
Full description #444

### CODE REVIEW

File: tests/index.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /* eslint-disable import/first */
-
 import './utils'
+/*
```
Comment (kof) [inline tests/index.js]: migrating plugins is wip

**Comment by kof** [comment]
ok @cssinjs/core this is ready to be reviewed!

File: src/types.js [hunk 2]
```diff
@@ -2,12 +2,17 @@
...
 export type ToCssOptions = {
   indent?: number
 }
 
-export type generateClassName = (str: string, rule: Rule) => string
+export type generateClassName = (rule: Rule) => string
+
+// TODO
+// Find a way to declare all types: Object|string|Array<Object>
+export type JssStyle = Object
```
Comment (iamstarkov) [inline src/types.js]: seems like its legit type def https://goo.gl/9RnYVu
screen shot 2017-04-05 at 10 57 07

Comment (kof) [inline src/types.js]: I narrowed down the problem in [this tweet](https://twitter.com/oleg008/status/844634629718102017)

https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoALgTwA4FMwDKWM+AvGAPIBGAVrgMboY74AiDMY5RmJAPgEEATkICGmADw8SAPlRgwqKAFcAdowCWcVWAAWo1QBMS03AAoAzsVwAuQtYCUYAN6oAvqiVrN2vQeO4wmKYZoYcduz0ME6uChpQYGZB4gB0GhbJIWFRDk5CuOjKQjrOHh4KXuroWjr6RiSh4WCR0S7yYPTaVmD56Fx+9YEi4o057fGJvXkFRTq97b2zAwGmo9HuQA

Comment (kof) [inline src/types.js]: Have to try phantom types.

Comment (iamstarkov) [inline src/types.js]:
```diff
/* @flow */
type Style = Object
type Decl = Style | Array<Style>
  
-function handleStyle(style: Style) {
+function handleStyle(style: Decl) {
}

function handleArray(decl: Decl) {
  if (Array.isArray(decl)) return {}
}
  
function handle(decl: Decl) {
  const ret = handleArray(decl)
  if (ret) return ret
  return handleStyle(decl)
}
```

Comment (kof) [inline src/types.js]: Sure, the pobl. is that handleStyle accepts only Style in that example, to be specific to our code base: RegularRule(name, style) where style is always an object. We don't expect arrays there. On the other hand FontFaceRule(selector, style) accepts style as object OR as array in order to support multiple [font faces at once](http://cssinjs.org/json-api?v=v6.5.0#font-face)

So the check `if (Array.isArray)` is local in FontFaceRule and RegularRule, but flow only can detect this if this check would be located in createRule() function which creates all the rules.

Comment (iamstarkov) [inline src/types.js]: ok

**Pull request merged**